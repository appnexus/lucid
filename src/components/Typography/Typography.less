@import (reference) '../../styles/variables.less';
@import (reference) '../../styles/mixins.less';

.@{prefix}-Typography {
	// general text
	&-p {
		font-family: @fontFamily;
	}
	// text for tables
	&-tabular {
		font-family: @font-family-tabular;
	}

	// headings
	&-h1,
	&-h2,
	&-h3 {
		margin: 0;
		font-weight: @font-weight-medium;
		text-rendering: optimizelegibility;
	}
	&-h1 {
		font-size: 24px;
		line-height: 28px;
	}
	&-h2 {
		font-size: 16px;
		line-height: 20px;
	}
	&-h3 {
		font-size: 12px;
		line-height: 16px;
		text-transform: uppercase;
		color: @color-neutral-7;
		letter-spacing: 0.8px;
	}

	// links
	&-a {
		font-weight: @font-weight-medium;
		color: @color-linkColor;
		text-decoration: none;
	}
	&-a:hover {
		color: @color-linkColorHover;
		text-decoration: underline;
	}

	// code and code-like
	&-code,
	&-pre,
	&-kbd,
	&-samp {
		// TODO: replace this with the variable from lucid when this issue is closed:
		// https://github.com/appnexus/lucid/issues/618
		font-family: Menlo, Monaco, 'Ubuntu Mono', Courier, monospace;
	}
}
