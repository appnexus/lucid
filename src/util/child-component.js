import React from 'react';
import _ from 'lodash';

/**
 * createChildComponent
 *
 * Returns a React Component which does not render anything, but has some useful static functions
 *
 * @return {ReactComponent} which renders `null` but exposes static helper functions
 */
export function createChildComponent (definition={}) {
	return React.createClass({
		propTypes: _.get(definition, 'propTypes', null),
		displayName: _.get(definition, 'displayName', null),
		render: () => null,
		statics: {

			/**
			 * getOwnChildren
			 *
			 * Returns all collective children of elements with this component type
			 *
			 * @param {Array} elements - usually `this.props.children` but can be any list of elements which
			 *	                          might have this type
			 * @return {Array} All the collective children of elements with this component type
			 */
			getOwnChildren(elements) {
				let component = this;
				let children = [];
				React.Children.forEach(elements, (element) => {
					if (element && element.type === component) {
						children = children.concat(React.Children.map(element.props.children, x => x));
					}
				});
				return children;
			},

			/**
			 * getAll
			 *
			 * Returns all elements with this component type
			 *
			 * @param {Array} elements - usually `this.props.children` but can be any list of elements which
			 *	                          might have this type
			 * @return {Array} elements with this component type
			 */
			getAll(elements) {
				let component = this;
				let elementsOfThisType = [];
				React.Children.forEach(elements, (element) => {
					if (element && element.type === component) {
						elementsOfThisType = elementsOfThisType.concat(element);
					}
				});
				return elementsOfThisType;
			},

			/**
			 * getAllProps
			 *
			 * Returns the `props` object of each element with this component type
			 *
			 * @param {Array} elements - usually `this.props.children` but can be any list of elements which
			 *	                          might have this type
			 * @return {Array} `props` objects of elements with this component type
			 */
			getAllProps(elements) {
				let component = this;
				return _.map(component.getAll(elements), 'props');
			},

			getAllAsProps(props, elements) {
				let component = this;
				return _.map(_.flatten([_.get(props, definition.propName, [])]), (childProp) => {
					if (_.isPlainObject(childProp)) {
						return _.assign({}, childProp, {children: childProp.children || childProp.text});
					} else {
						return {
							children: childProp
						};
					}
				}).concat(component.getAllProps(elements));
			}
		}
	});
}
