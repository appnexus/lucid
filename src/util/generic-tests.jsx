import sinon from 'sinon';
import React from 'react';
import { mount, shallow } from 'enzyme';
import assert from 'assert';
import describeWithDOM  from './describe-with-dom';
import _ from 'lodash';
import * as lucid from '../index';

// Common tests for all our components
export function common(Component, {
	getDefaultProps = _.noop,
	selector = null,
	exemptFunctionProps = [],
} = {}) {

	function generateDefaultProps(props={}) {
		return _.assign({}, getDefaultProps(), props);
	}

	describe('[common]', () => {
		it('should have a `displayName` defined', () => {
			assert(Component.displayName);
		});

		it('should pass through styles to the root element', () => {
			let style = {
				backgroundColor: '#f0f',
			};
			const wrapper = shallow(<Component {...generateDefaultProps()} style={style}/>);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const rootStyle = rootWrapper.prop('style');
			assert(_.every(style, (val, key) => val === rootStyle[key]), 'root style must contain passed styles');
		});

		it('should pass through `className`', () => {
			const expectedClass = 'rAnDoM';
			const wrapper = shallow(<Component {...generateDefaultProps()} className={expectedClass}/>);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(_.includes(classNames, expectedClass), `'${classNames}' should include '${expectedClass}'`);
		});

		it('should have an application scoped base class', () => {
			const expectedClass = 'lucid-' + Component.displayName;
			const wrapper = shallow(<Component {...generateDefaultProps()} />);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(_.includes(classNames, expectedClass), `'${classNames}' should include '${Component.displayName}'`);
		});

		it('should have only application scoped classes', () => {
			const wrapper = shallow(<Component {...generateDefaultProps()} />);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const parentClasses = rootWrapper.prop('className').split(' ');
			const childrenClasses = rootWrapper.children().reduce((acc, node) => {
				if (!node.prop('className')) {
					return acc;
				}

				return acc.concat(node.prop('className').split(' '))
			}, []);

			const allClasses = parentClasses.concat(childrenClasses);

			_.forEach(allClasses, className => {
				assert(_.includes(className, `lucid-${Component.displayName}`), `${className} must be scoped`);
			});
		});

		it('should only use onX convention for function proptypes', () => {

			_.forEach(Component.propTypes, (value, key) => {
				// See the following tests from the React source code to figure out how
				// this works: https://github.com/facebook/react/blob/v0.14.7/src/isomorphic/classic/types/__tests__/ReactPropTypes-test.js
				let props = {};
				props[key] = _.noop;

				// Here we test the validation against a noop function, if it doesn't
				// have an error, then it's probably a function. The other possible is
				// that it's an `any`, so we'll test for that below
				const isProbablyFunction = !(value(props, key) instanceof Error);

				// We'll also test out a string to see if that valid
				props[key] = '';

				const isAny = isProbablyFunction && !(value(props, key) instanceof Error);

				// If it's probably a function, and it's not `any`, then we make sure
				// it starts with `on`

				if (isProbablyFunction) {
					assert(isAny || _.startsWith(key, 'on') || _.includes(exemptFunctionProps, key), `${key} must follow onX convention`);
				}

			});

		});

		// Only run this test if it's a public component
		if (!Component._lucidIsPrivate) {
			it('should be available as an exported module from index.js', () => {
				assert(lucid[Component.displayName]);
			});
		}
	});
}

// Common tests for all our icon components
export function icons(Component) {
	describeWithDOM('[icon]', () => {
		it('should pass through isBadge prop to underlying Icon component', () => {
			const wrapper = mount(<Component isBadge={true} />);
			const classNames = wrapper.find('svg').prop('className').split(' ');
			const targetClassName = 'lucid-Icon-is-badge';
			assert(_.includes(classNames, targetClassName), `'${classNames}' should include '${targetClassName}'`);
		});
	});
}

// Common tests for all control components
export function controls(Component, { callbackName, controlSelector , eventType }) {
	// Use DOM tests here since some of our controls use dom events under the hood
	describeWithDOM('[control]', () => {
		it('should callback with `event` and `props`', () => {
			const expectedSpecialProp = 32;
			const props = {
				specialProp: expectedSpecialProp,
				[callbackName]: sinon.spy(),
			};
			const wrapper = mount(<Component {...props} />);

			wrapper.find(controlSelector).simulate(eventType);

			// Last argument should be an object with `uniqueId` and `event`
			const { props: { specialProp }, event } = _.last(props[callbackName].args[0]);

			assert(event, 'missing event');
			assert.equal(specialProp, expectedSpecialProp, 'incorrect or missing specialProp');
		});
	});
}
