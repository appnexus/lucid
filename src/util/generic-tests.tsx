import sinon from 'sinon';
import { parse } from 'path';
import React from 'react';
import PropTypes from 'prop-types';
import { mount, shallow, render } from 'enzyme';
import assert from 'assert';
import _, { each, omit, keys, includes } from 'lodash';
import glob from 'glob';
import { addons, mockChannel } from '@storybook/addons';
import timekeeper from 'timekeeper';

addons.setChannel(mockChannel());

interface ICommonConfig {
	getDefaultProps?: () => {};
	exemptFunctionProps?: string[];
	exemptChildComponents?: string[];
	selectRoot?: any;
	noExport?: boolean;
}

// Common tests for all our components
export function common(Component: any, config: ICommonConfig = {} as any) {
	const {
		getDefaultProps = _.constant({}),
		exemptFunctionProps = [],
		exemptChildComponents = [],
		selectRoot = _.identity,
		noExport = false,
	} = config;

	function generateDefaultProps(props = {}) {
		return _.assign({}, getDefaultProps(), props);
	}

	describe('[common]', () => {
		if (!Component) {
			throw new Error('An undefined component was passed to generic tests.');
		}

		if (Component._isLucidHybridComponent) {
			throw new Error(
				`You're trying to run generic tests on a hybrid component which is bad and won't work and will make you cry. Check your spec files for ${Component.displayName} and import the raw component instead of the hybrid version.`
			);
		}

		it('should have a `displayName` defined', () => {
			assert(Component.displayName);
		});

		it('should pass through styles to the root element', () => {
			const style = {
				backgroundColor: '#f0f',
			};
			const wrapper = shallow(
				<Component {...generateDefaultProps()} style={style} />,
				{ disableLifecycleMethods: true }
			);

			const rootWrapper = selectRoot(wrapper).first();
			const rootStyle = rootWrapper.prop('style');
			assert(
				_.every(style, (val, key) => val === rootStyle[key]),
				'root style must contain passed styles'
			);
		});

		it('should pass through `className`', () => {
			const expectedClass = 'rAnDoM';
			const wrapper = shallow(
				<Component {...generateDefaultProps()} className={expectedClass} />,
				{ disableLifecycleMethods: true }
			);
			const rootWrapper = selectRoot(wrapper).first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(
				_.includes(classNames, expectedClass),
				`'${classNames}' should include '${expectedClass}'`
			);
		});

		it('should have an application scoped base class', () => {
			const expectedClass = 'lucid-' + Component.displayName;
			const wrapper = shallow(<Component {...generateDefaultProps()} />, {
				disableLifecycleMethods: true,
			});
			const rootWrapper = selectRoot(wrapper).first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(
				_.includes(classNames, expectedClass),
				`'${classNames}' should include '${Component.displayName}'`
			);
		});

		it('should have only application scoped classes', () => {
			const wrapper = shallow(<Component {...generateDefaultProps()} />, {
				disableLifecycleMethods: true,
			});
			const rootWrapper = selectRoot(wrapper).first();
			const parentClasses = rootWrapper.prop('className').split(' ');
			const childrenClasses = rootWrapper.children().reduce((acc, node) => {
				if (!node.prop('className')) {
					return acc;
				}

				return acc.concat(node.prop('className').split(' '));
			}, []);

			const allClasses = parentClasses.concat(childrenClasses);

			_.forEach(allClasses, (className) => {
				assert(
					_.includes(className, `lucid-${Component.displayName}`),
					`${className} must be scoped`
				);
			});
		});

		describe('function propTypes', () => {
			const funcProps = _.pickBy(
				Component.propTypes,
				(propType) => propType === PropTypes.func
			);

			_.forEach(funcProps, (propType, propName) => {
				it(`${propName} should only use onX convention for function proptypes`, () => {
					assert(
						_.startsWith(propName, 'on') ||
							_.includes(exemptFunctionProps, propName),
						`${propName} must follow onX convention`
					);
				});
			});
		});

		describe('child components', () => {
			// Child components are all function types which start with a capital letter
			const childComponents = _.pickBy(Component, (value, key) => {
				return /^[A-Z]/.test(key) && _.isFunction(value);
			});

			describe('propNames in propTypes', () => {
				_.flow(
					(x) => _.map(x, 'propName'),
					(x) => _.compact(x),
					(x) => _.flatMap(x, _.castArray),
					(x) =>
						_.reject(x, (propName) =>
							_.includes(exemptChildComponents, propName)
						),
					(x) =>
						_.forEach(x, (propName) => {
							it(`should include ${propName} in propTypes`, () => {
								assert(
									Component.propTypes[propName],
									`must include ${propName} in propTypes`
								);
							});
						})
				)(childComponents);
			});
		});

		describe('example testing', () => {
			const fileNames = glob.sync(
				`./src/components/**/${Component.displayName}/*.stories.@(j|t)sx`
			);
			each(fileNames, (path) => {
				const lib = require('../../' + path.replace('.tsx', ''));

				each(omit(lib, ['default']), (Story, name) => {
					it(`should match snapshot(s) for ${name}`, () => {
						let result;
						try {
							result = render(<Story {...Story.args} />, {
								disableLifecycleMethods: true,
							});
							expect(result).toMatchSnapshot();
						} catch (err) {
							expect(err).toMatchSnapshot();
						}
					});
				});
			});

			// Support for older examples
			const exampleFileNames = glob.sync(
				`./src/components/**/${Component.displayName}/examples/*.@(j|t)sx`
			);
			_.each(exampleFileNames, (path) => {
				const lib = require('../../' + path.replace('.tsx', ''));

				const Example = lib.default;

				const title = parse(path).name;
				it(`should match snapshot(s) for ${title}`, () => {
					const shallowExample = shallow(<Example />, {
						disableLifecycleMethods: true,
					});

					// If the root of the example is an instance of the Component under test, snapshot it.
					// Otherwise, look under the root for instances of the Component and snapshot those.
					if (shallowExample.is(Component.displayName)) {
						expect(
							shallow(<Component {...shallowExample.props()} />, {
								disableLifecycleMethods: true,
							})
						).toMatchSnapshot();
					} else {
						shallowExample.find(Component.displayName).forEach((example) => {
							expect(
								shallow(<Component {...example.props()} />, {
									disableLifecycleMethods: true,
								})
							).toMatchSnapshot();
						});
					}
				});
			});
		});
	});
}

// Common tests for all our icon components

interface IIconConfig {
	includeInitialState: boolean;
}

export function icons(Component: any, config: IIconConfig = {} as any) {
	// The default expectation is for every Icon to omit `initailState`,
	// if it is passed through to the underlying element
	const { includeInitialState = false } = config;

	describe('[icon]', () => {
		it('should almost always omit the `initialState` key', () => {
			const wrapper = shallow(<Component initialState={{ testState: true }} />);
			const rootProps = keys(wrapper.first().props());

			expect(includes(rootProps, 'initialState')).toBe(includeInitialState);
		});

		it('should add the correct class for isClickable', () => {
			const wrapper = mount(<Component isClickable={true} />);
			const targetClassName = 'lucid-Icon-is-clickable';
			assert(
				wrapper.find('svg').hasClass(targetClassName),
				`Missing '${targetClassName}' class`
			);
		});
	});
}

// Common tests for all control components
export function controls(
	Component: any,
	{ callbackName, controlSelector, eventType, additionalProps = {} }: any
) {
	// Use DOM tests here since some of our controls use dom events under the hood
	describe('[control]', () => {
		it('should callback with `event` and `props`', () => {
			const expectedSpecialProp = 32;
			const props = {
				specialprop: expectedSpecialProp,
				[callbackName]: sinon.spy(),
				...additionalProps,
			};
			const wrapper = mount(<Component {...props} />);

			wrapper.find(controlSelector).first().simulate(eventType);

			// Last argument should be an object with `uniqueId` and `event`
			const {
				props: { specialprop },
				event,
			}: any = _.last(props[callbackName].args[0]);

			assert(event, 'missing event');
			assert.equal(
				specialprop,
				expectedSpecialProp,
				'incorrect or missing specialProp'
			);
		});
	});
}

// Common tests for all Functional Components
//
// These tests are intended to help us make sure our FCs are shaped corrected.
// They are necessary because there isn't a perfect way to get the defaultProps
// to be factored in correctly yet with React/TypeScript:
// https://github.com/DefinitelyTyped/DefinitelyTyped/issues/30695#issuecomment-474780159
export function functionalComponents(FC: any) {
	// Use DOM tests here since some of our controls use dom events under the hood
	describe('[functionalComponent]', () => {
		it('should have the correct `peek` properties', () => {
			expect(FC.propName === undefined || typeof FC.propName === 'string').toBe(
				true
			);
			expect(
				FC._isPrivate === undefined || typeof FC._isPrivate === 'boolean'
			).toBe(true);

			expect(typeof FC.peek).toBe('object');
			expect(typeof FC.peek.description).toBe('string');
			expect(
				FC.peek.extend === undefined || typeof FC.peek.extend === 'string'
			).toBe(true);
			expect(
				FC.peek.extend === undefined || typeof FC.peek.extend === 'string'
			).toBe(true);
			expect(
				FC.peek.categories === undefined || Array.isArray(FC.peek.categories)
			).toBe(true);
			expect(
				FC.peek.madeFrom === undefined || Array.isArray(FC.peek.madeFrom)
			).toBe(true);
		});
	});
}

export const mockDate = (dateString: string) => {
	timekeeper.freeze(new Date(dateString));
};
