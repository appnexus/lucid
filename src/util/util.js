import _ from 'lodash';
import React from 'react';

export const isDevMode = (function isReactInDev() {
	try {
		React.createClass({});
	} catch(e) {
		if (e.message.indexOf('render') !== 0) {
			return true;  // A nice, specific error message
		}
	}

	return false;  // should never happen, but play it safe.
})();

export const isNode = typeof process === 'object' && process.title === 'node';

export const logger = (function() {
	return checkIsDev() ? {
		log,
		warn,
		error
	} : {
		log: _.noop,
		warn: _.noop,
		error: _.noop
	};
})();

export function checkIsDev() {
	return isDevMode &&
		typeof window !== 'undefined' &&
		!isNode &&
		typeof console !== 'undefined';
}

function log(...args) {

	console.log(...args);

	try {
		// --- Welcome to debugging Lucid ---
		// This error was thrown as a convenience so that you can use this
		// stack to find the callsite that caused this warning to fire.
		throw new Error(args[0]);
	} catch (x) {}

}

function warn(...args) {

	console.warn(...args);

	try {
		// --- Welcome to debugging Lucid ---
		// This error was thrown as a convenience so that you can use this
		// stack to find the callsite that caused this warning to fire.
		throw new Error(args[0]);
	} catch (x) {}

}

function error(...args) {

	console.error(...args);

	try {
		// --- Welcome to debugging Lucid ---
		// This error was thrown as a convenience so that you can use this
		// stack to find the callsite that caused this warning to fire.
		throw new Error(args[0]);
	} catch (x) {}

}
