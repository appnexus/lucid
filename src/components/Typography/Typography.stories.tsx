import React from 'react';
//import createClass from 'create-react-class';
import { Story, Meta } from '@storybook/react';

import HelpIcon from './../Icon/HelpIcon/HelpIcon';
import ToolTip from './../ToolTip/ToolTip';
import Typography, { ITypographyProps } from './Typography';

export default {
	title: 'Text/Typography',
	component: Typography,
	parameters: {
		docs: {
			description: {
				component: Typography.peek.description,
			},
		},
	},
} as Meta;

/* Variants */
export const Variants: Story<ITypographyProps> = (args) => {
	const { Title, Target } = ToolTip;

	return (
		<section>
			<Typography {...args} variant='h1'>
				h1. Heading
			</Typography>
			<br />
			<Typography {...args} variant='h2'>
				h2. Heading
			</Typography>
			<br />
			<Typography {...args} variant='h3'>
				h3. Heading
			</Typography>
			<br />
			<Typography {...args} variant='p'>
				p. paragraph text
			</Typography>
			<br />
			<Typography {...args} variant='tabular'>
				tabular. text for tables
			</Typography>
			<br />
			<Typography {...args} variant='a'>
				a. link text
			</Typography>
			<br />
			<Typography {...args} variant='pre'>
				pre. preformatted text
			</Typography>
			<br />
			<Typography {...args} variant='code'>
				code. snazzy code
			</Typography>
			<br />
			<Typography {...args} variant='kbd'>
				kbd. keyboard inputs
			</Typography>
			<br />
			<Typography {...args} variant='samp'>
				samp. sample code outputs
			</Typography>
			<br />
			<Typography {...args} variant='span'>
				span. sample code outputs with help bubble
				<div className='HelpBubble'>
					<ToolTip alignment={'start'} direction={'right'}>
						<Title className='HelpBubble-title'>look at me</Title>
						<Target>
							<HelpIcon size={12} />
						</Target>
					</ToolTip>
				</div>
			</Typography>{' '}
			<br />
		</section>
	);
};

/* Nested */
export const Nested: Story<ITypographyProps> = (args) => {
	return (
		<section>
			<Typography {...args} variant='h1'>
				Title
			</Typography>
			<Typography {...args} variant='h2'>
				subtitle
			</Typography>
			<Typography {...args} variant='p'>
				Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod
				tempor incididunt ut labore et dolore magna aliqua. Pencil mustachioed
				graeme souness, mustachioed waxy gurn rugged el snort graeme souness
				burt reynolds educated village people boogie nights pencil man of the
				year 1986, pencil village people graeme souness rugged boogie nights
				mouthbrow educated burt reynolds waxy gurn man of the year 1986 des
				lynam charming villain mustachioed cardinal richelieu el snort. Borat
				d’artagnan hungarian testosterone trophy frontiersman hairy kiss.
				beefeater chevron, hairy kiss. cardinal richelieu groucho-a-like
				testosterone trophy d’artagnan dick van dyke rugged chevron Droopy
				frontiersman borat beefeater hungarian chevron, cardinal richelieu
				d’artagnan beefeater el snort Droopy dodgy uncle clive groucho-a-like
				hairy kiss. chevron frontiersman chevron dick van dyke walrus leader of
				men rugged borat bad guy testosterone trophy groucho-a-like alpha trion
				hungarian.
				<br />
				<br />
				<Typography {...args} variant='a' style={{ display: 'block' }}>
					Like!
				</Typography>
				<br />
				<Typography {...args} variant='a' style={{ display: 'block' }}>
					Subscribe!
				</Typography>
				<br />
				<Typography {...args} variant='h3'>
					Code Example
				</Typography>
				<Typography {...args} variant='code'>
					npm install all-the-things
				</Typography>
			</Typography>
		</section>
	);
};
