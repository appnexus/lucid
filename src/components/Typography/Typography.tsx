import _ from 'lodash';
import React from 'react';
import PropTypes from 'react-peek/prop-types';
import { lucidClassNames } from '../../util/style-helpers';
import { omitProps, StandardProps } from '../../util/component-types';

const cx = lucidClassNames.bind('&-Typography');

const { node, string, oneOf } = PropTypes;

export enum ElementTypes {
	p = 'p',
	tabular = 'p',
	h1 = 'h1',
	h2 = 'h2',
	h3 = 'h3',
	a = 'a',
	pre = 'pre',
	code = 'code',
	kbd = 'kbd',
	samp = 'samp',
	span = 'span',
}

export interface ITypographyProps
	extends StandardProps,
		React.DetailedHTMLProps<React.HTMLAttributes<HTMLElement>, HTMLElement> {
	/** This prop defines the type of text that will be displayed. It may be an actual HTML element or something with extra semantic meaning. */
	variant: keyof typeof ElementTypes;
}

export const Typography = (props: ITypographyProps) => {
	const { children, className, variant, ...passThroughs } = props;
	const Element = ElementTypes[variant ? variant : 'p'];

	return React.createElement(
		Element,
		{
			...omitProps<ITypographyProps>(
				passThroughs,
				undefined,
				_.keys(Typography.propTypes)
			),
			className: cx('&', `&-${variant}`, className),
		},
		children
	);
};

Typography.displayName = 'Typography';
Typography.peek = {
	description: `
		A general component for several types of textual HTML elements.
	`,
	categories: ['text'],
};
Typography.propTypes = {
	children: node`
		Can contain elements or nested \`Typography\` components.
	`,

	className: string`
		Appended to the component-specific class names set on the root element.
	`,

	variant: oneOf(_.keys(ElementTypes)).isRequired`
		This prop defines the type of text that will be displayed.
		It may be an actual HTML element or something with extra semantic meaning.
	`,
};

export default Typography;
