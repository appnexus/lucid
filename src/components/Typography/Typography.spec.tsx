import { common } from '../../util/generic-tests';
import Typography from './Typography';

describe('Typography', () => {
	common(Typography, {
		getDefaultProps: () => ({
			variant: 'p',
		}),
	});
});
