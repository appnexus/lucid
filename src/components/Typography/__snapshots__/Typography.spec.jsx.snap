// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography [common] example testing should match snapshot(s) for 01.variants 1`] = `
<h1
  className="lucid-Typography lucid-Typography-h1"
>
  h1. Heading
</h1>
`;

exports[`Typography [common] example testing should match snapshot(s) for 01.variants 2`] = `
<h2
  className="lucid-Typography lucid-Typography-h2"
>
  h2. Heading
</h2>
`;

exports[`Typography [common] example testing should match snapshot(s) for 01.variants 3`] = `
<h3
  className="lucid-Typography lucid-Typography-h3"
>
  h3. Heading
</h3>
`;

exports[`Typography [common] example testing should match snapshot(s) for 01.variants 4`] = `
<p
  className="lucid-Typography lucid-Typography-p"
>
  p. paragraph text
</p>
`;

exports[`Typography [common] example testing should match snapshot(s) for 01.variants 5`] = `
<p
  className="lucid-Typography lucid-Typography-tabular"
>
  tabular. text for tables
</p>
`;

exports[`Typography [common] example testing should match snapshot(s) for 01.variants 6`] = `
<a
  className="lucid-Typography lucid-Typography-a"
>
  a. link text
</a>
`;

exports[`Typography [common] example testing should match snapshot(s) for 01.variants 7`] = `
<pre
  className="lucid-Typography lucid-Typography-pre"
>
  pre. preformatted text
</pre>
`;

exports[`Typography [common] example testing should match snapshot(s) for 01.variants 8`] = `
<code
  className="lucid-Typography lucid-Typography-code"
>
  code. snazzy code
</code>
`;

exports[`Typography [common] example testing should match snapshot(s) for 01.variants 9`] = `
<kbd
  className="lucid-Typography lucid-Typography-kbd"
>
  kbd. keyboard inputs
</kbd>
`;

exports[`Typography [common] example testing should match snapshot(s) for 01.variants 10`] = `
<samp
  className="lucid-Typography lucid-Typography-samp"
>
  samp. sample code outputs
</samp>
`;

exports[`Typography [common] example testing should match snapshot(s) for 01.variants 11`] = `
<span
  className="lucid-Typography lucid-Typography-span"
>
  span. sample code outputs with help bubble
  <div
    className="HelpBubble"
  >
    <ToolTip
      alignment="start"
      direction="right"
      name="I am a div under typography"
    >
      <ToolTip.Title
        className="HelpBubble-title"
      >
        look at me
      </ToolTip.Title>
      <ToolTip.Target>
        <HelpIcon
          aspectRatio="xMidYMid meet"
          color="primary"
          isClickable={false}
          isDisabled={false}
          onClick={[Function]}
          onSelect={[Function]}
          size={12}
          viewBox="0 0 16 16"
        />
      </ToolTip.Target>
    </ToolTip>
  </div>
</span>
`;

exports[`Typography [common] example testing should match snapshot(s) for 02.nested 1`] = `
<h1
  className="lucid-Typography lucid-Typography-h1"
>
  Title
</h1>
`;

exports[`Typography [common] example testing should match snapshot(s) for 02.nested 2`] = `
<h2
  className="lucid-Typography lucid-Typography-h2"
>
  subtitle
</h2>
`;

exports[`Typography [common] example testing should match snapshot(s) for 02.nested 3`] = `
<p
  className="lucid-Typography lucid-Typography-p"
>
  Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Pencil mustachioed graeme souness, mustachioed waxy gurn rugged el snort graeme souness burt reynolds educated village people boogie nights pencil man of the year 1986, pencil village people graeme souness rugged boogie nights mouthbrow educated burt reynolds waxy gurn man of the year 1986 des lynam charming villain mustachioed cardinal richelieu el snort. Borat d’artagnan hungarian testosterone trophy frontiersman hairy kiss. beefeater chevron, hairy kiss. cardinal richelieu groucho-a-like testosterone trophy d’artagnan dick van dyke rugged chevron Droopy frontiersman borat beefeater hungarian chevron, cardinal richelieu d’artagnan beefeater el snort Droopy dodgy uncle clive groucho-a-like hairy kiss. chevron frontiersman chevron dick van dyke walrus leader of men rugged borat bad guy testosterone trophy groucho-a-like alpha trion hungarian.
  <br />
  <br />
  <Typography
    style={
      Object {
        "display": "block",
      }
    }
    variant="a"
  >
    Like!
  </Typography>
  <Typography
    style={
      Object {
        "display": "block",
      }
    }
    variant="a"
  >
    Subscribe!
  </Typography>
  <br />
  <Typography
    variant="h3"
  >
    Code Example
  </Typography>
  <Typography
    variant="code"
  >
    npm install all-the-things
  </Typography>
</p>
`;

exports[`Typography [common] example testing should match snapshot(s) for 02.nested 4`] = `
<a
  className="lucid-Typography lucid-Typography-a"
  style={
    Object {
      "display": "block",
    }
  }
>
  Like!
</a>
`;

exports[`Typography [common] example testing should match snapshot(s) for 02.nested 5`] = `
<a
  className="lucid-Typography lucid-Typography-a"
  style={
    Object {
      "display": "block",
    }
  }
>
  Subscribe!
</a>
`;

exports[`Typography [common] example testing should match snapshot(s) for 02.nested 6`] = `
<h3
  className="lucid-Typography lucid-Typography-h3"
>
  Code Example
</h3>
`;

exports[`Typography [common] example testing should match snapshot(s) for 02.nested 7`] = `
<code
  className="lucid-Typography lucid-Typography-code"
>
  npm install all-the-things
</code>
`;
